# The App Store and Google Play Ad Revenue Breakdown

We are looking to find the most profitable markets for mobile applications in the App Store and Google play markets. Look at various data sets, we can take a data-driven approach to developing new applications with high profit returns.

The goal of this project is to breakdown the markets to find the best approach to building new applications. As a company that only builds free apps, ads play a huge part in the decision-making process, as it is the source of income. Analyzing user bases and other various indicators will help influence the direction of development and marketing.

## Datasets

For our analysis, we are using two datasets:

[A Google Play dataset](https://www.kaggle.com/lava18/google-play-store-apps/home) featuring 10,000 apps from the Google Play Store, collected in August 2018.

[An App Store dataset](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home) featuring 7,000 app form the App Store, collected in July 2017.



In [1]:
from csv import reader

# Google Play dataset
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
google_data = list(read_file)
android_header = google_data[0]
android = google_data[1:]

# App Store dataset
opened_file = open('AppleStore.csv')
read_file = reader(opened_file)
apple_data = list(read_file)
ios_header = apple_data[0]
ios = apple_data[1:]

To start, we gather the datasets from both the Google Play Store and the App Store. To make things easier to reference, we separate the header from the rest of the data.

Let's look at the Play Store dataset.

In [2]:
def explore_data(dataset, first = 0, last = 5):
    data_section = dataset[first:last]
    for row in data_section:
        print(row)
        print('\n')
    print('Number of rows: ', len(dataset))
    print('Number of columns: ', len(dataset[0]))

print(android_header)
print('\n')
explore_data(android)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Eve

As you can see, there are 13 columns we can use for the dataset as well as 10841 rows of data. Of these columns, some of the most useful will be `'App'`, `'Category'`, `'Reviews'`, `'Installs'`, `'Type'`, `'Price'`, and `'Genres'`.

Next, we will look at the App Store data set.

In [3]:
print(ios_header)
print('\n')
explore_data(ios)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of rows:  7197
Number of columns:  16


The App Store has 16 columns to use as well as 7197 rows of data to use. Some of the colummns we will use are `'track_name'`, `'currency'`, `'price'`, `'rating_count_tot'`, `'rating_count_ver'`, and `'prime_genre'`.

These columns are not self-explanitary, so to see what these columns do specifically, check out the [documentation](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home).

## Data Cleaning

Before we get to far, we want to clean up the data before running. First, there is a row (#10472) in the Google Play Store that is incorrectly formatted, as referenced in the [discussion here](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015). Let's look at what is going on.

In [4]:
print(android[10472]) # incorrect row
print('\n')
print(android_header) # header
print('\n')
print(android[0]) # correct row

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


We can see that the second column, Category, does not have an entry in row 10472. We will have to remove from our dataset.

In [5]:
print('Android columns before removal- ', len(android))
del android[10472] # removes from dataset
print('Android columns after removal- ', len(android))

Android columns before removal-  10841
Android columns after removal-  10840


## Duplicate Data

In the Google Play dataset, there are many instances of duplicate data, which needs to be handled.

In [6]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


As you can see, there are four entries for Instagram. There are many more instances of this throughout the dataset.

In [7]:
duplicate_apps = []
unique_apps = []

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate apps- ', len(duplicate_apps))
print('\n')
print('Some examples- ', duplicate_apps[:15])

Number of duplicate apps-  1181


Some examples-  ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


As we can see, there is 1181 duplicate apps in our dataset. We will have to figure out a way to remove these from our dataset. 

If we look at the Instagram data above, we can see that the only difference between each row is the number of reviews. This will tell us that the data for each row was taken at different times. Instead of deleting rows at random, we can look at the highest humber of ratings to help guide us towards the most accurate row to use.

We will want to create a new dictionary where each entry is has a unique app name which has the highest amount of reviews, then use that dictionary as our new dataset.

First, we need to figure out how many rows we should see once we get rid of the duplicates.

In [8]:
print('Expected length- ', len(android) - len(duplicate_apps))

Expected length-  9659


We are expecting to see 9659 rows in our new dataset.

Now, we will make a new dictionary for each on the unique apps with the maximum amount of reviews.

In [9]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print(len(reviews_max))

9659


As you can see, the length of this new dictionary is the same length as we expected from our previous calculations. We went through the android dataset and stored the name and the review amounts. If the name was already in the dictionary and the reviews that were there are less than the current row, we want to add the new value to the dictionary. Also, we want to add any name that is not yet added to the dictionary as well. So when a name is not in the dictionary, we will just simply add it. If it is already there, we want to test to see if the current value for the row is larger than the previous result. If it is, override the current selection.



In [10]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews=float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)
        
print('Amount in android_clean- ', len(android_clean))

Amount in android_clean-  9659


To clean this database, we are going to create some containers, one for the clean database and one to compare the names to ones that are already added to the database.

We are then going to take the data in android and check the names and reviews. If the current review number is the same as what we have in the reviews_max dictionary, we know that it is the largest number of reviews. We also want to check to make sure that the name is not already in the already_added list. Once the criteria is met, we append the full row to the android_clean and the name to the already_added list.

We can run the explore data function to give us a quick look at this new dataset.

In [11]:
print(android_header)
print('\n')
explore_data(android_clean)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Every

As we can see, we have the same number of rows as we had before.

## Getting rid of non-English apps

As we are a company that deals in the English speaking market, we will need to remove apps that do not fit into our demographic. Let's look at a couple of examples of these non-English apps.

In [12]:
print(ios[813][1])
print(ios[6731][1])
print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜
中国語 AQリスニング
لعبة تقدر تربح DZ


As you can see, these apps would not be useful for our purposes.

To get rid of these apps, we will look for characters that are not in English. To do this, we will use the [ASCII](https://en.wikipedia.org/wiki/ASCII) system.

In [13]:
print(ord('奇'))
print(ord('a'))

22855
97


In the ASCII system, the numbers 0 to 127 are highly used English characters. Everything else is reserved for other things. We can then assume that anything without an English character is not what we want to include in out dataset.

In [14]:
def english_check(string):
    for character in string:
        if ord(character) > 127:
            return False
        
    return True

print(ios[0][1]," - ", english_check(ios[0][1]))
print(ios[813][1]," - ", english_check(ios[813][1]))
print('Instachat 😜 - ',english_check('Instachat 😜'))
print('Docs To Go™ Free Office Suite - ', english_check('Docs To Go™ Free Office Suite'))

Facebook  -  True
爱奇艺PPS -《欢乐颂2》电视剧热播  -  False
Instachat 😜 -  False
Docs To Go™ Free Office Suite -  False


Let's look at what is happening. We created a function called english_check to help break down each character in a title of an app. We pass in a string (in this case the name of the app) and go through each character and check if it is within the ASCII limit of 0 to 127 for the English language. If it is over that limit, it returns false. Otherwise, it returns true once it scans the entire string. Facebook returns true, while the other returns false. However, our other examples will also fall outside the ASCII range, even though it really should fall within. We would end up losing some data that we would actually want. Let's revise the function to help combat the issue.

In [17]:
def english_check(string):
    bad_character = 0
    
    for character in string:
        if ord(character) > 127:
            bad_character += 1
        
    if bad_character > 3:
        return False
    else:
        return True

print(ios[0][1]," - ", english_check(ios[0][1]))
print(ios[813][1]," - ", english_check(ios[813][1]))
print('Instachat 😜 - ', english_check('Instachat 😜'))
print('Docs To Go™ Free Office Suite - ', english_check('Docs To Go™ Free Office Suite'))

Facebook  -  True
爱奇艺PPS -《欢乐颂2》电视剧热播  -  False
Instachat 😜 -  True
Docs To Go™ Free Office Suite -  True


We modified the function to essentially add up the total of bad characters throughout the title. If there are more than 3, it is considered a bad string. There might still be some apps that fall through the cracks, but it will be close enough for our use.

Now let's cut out those non-English apps.

In [21]:
android_english = []
ios_english = []

for app in android_clean:
    name = app[0]
    checker = english_check(name)
    
    if checker:
        android_english.append(app)
        
for app in ios:
    name = app[1]
    checker = english_check(name)
    
    if checker:
        ios_english.append(app)
        
explore_data(android_english)
print('\n')
explore_data(ios_english)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


Number of rows:  9614
Number of columns:  13


['284882215', 'Facebook', '389879808', 'USD'